### PR TITLE
add file modification event

### PR DESCRIPTION
**Pull request by roikol**
## Initial Checklist

- [x] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [ ] If part of an EPIC, PR git log contains EPIC number.
- [ ] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

```
Author: RoiKol <roi.kol@aquasec.com>
Date:   Thu Jan 19 18:54:19 2023 +0200

    events: add file_modification event
```

issue: #2728  

the `file_modification` event is derived from two internal events: `file_update_time` and `file_modified`.

the idea of `file_update_time` and `file_modified` is to hook on the points where ctime of a file in the system is chaged.

note:
there is a difference between kernel versions regarding the relations of these two kernel functions ===>

kernel ≥ 6:
__file_update_time() updates ctime.
file_update_time() calls __file_update_time()
file_modified() calls __file_update_time()

5.3 ≤ kernel < 6:
update_time() updates ctime.
file_update_time() calls update_time()
file_modified() calls file_update_time()

4.18 ≤ kernel < 5.3
update_time() updates ctime.
file_update_time() calls update_time()
(file_modified() doesn’t exist)

<====

therefore in the event of `file_modified` the probes are set to `required: false`.

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [ ] Code refactor (code improvement and/or code removal)
- [x] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

Reproduce the test by running:

- command 01: `./dist/tracee-ebpf -t e=file_modification`

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [x] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

File: pkg/events/events.go
Comment (NDStrahilevitz) [inline pkg/events/events.go:6498]: If you go by the approach suggested you wouldn't need this as an event.

File: pkg/events/events.go
Comment (NDStrahilevitz) [inline pkg/events/events.go:6515]: ditto

File: pkg/events/events.go
Comment (NDStrahilevitz) [inline pkg/events/events.go:6532]: Then the probe requirements go here directly.

**Review by NDStrahilevitz — CHANGES_REQUESTED** [review]
I think using the event derivation is redundant here as there is no logic done in the derivation, you may as well just submit the ID of `FILE_MODIFICATION` in the eBPF code.

And once we have feature detection I guess this can be optimized to have a "switched" probe or something.

**Review by NDStrahilevitz — APPROVED** [review]
LGTM @yanivagman if you also want to take a look before merging

**Comment by roikol** [comment]
thanks @NDStrahilevitz !

i've updated the PR to only output one event of file modification between open and close by a process - in order to avoid a high number of events (if a process calls a lot of `write`s for example)

**Review by NDStrahilevitz — APPROVED** [review]
LGTM, nice work with the new e2e. We should start adding existing events there. I believe we should wait with merging until the release is out.

**Review by NDStrahilevitz — CHANGES_REQUESTED** [review]
This needs a rebase to realign with the bpf code split.
I would love merging this asap so we can start adding new (and existing events) to these e2es.

**Comment by rafaeldtinoco** [comment]
@NDStrahilevitz I want to review before this is merged.

**Review by NDStrahilevitz — APPROVED** [review]
LGTM, waiting for @rafaeldtinoco

**Review by yanivagman — CHANGES_REQUESTED** [review]

File: pkg/ebpf/c/tracee.bpf.c
Comment (roikol) [reply pkg/ebpf/c/tracee.bpf.c:3872]: when i search the key in the map, i get a `struct file` from the `file_update_time() / file_modified()` events. 
i could seach backwards for the relevant FD in the process FD table, but it seems to me that device+inode is more efficient here

**Comment by rafaeldtinoco** [comment]
This PR should hold until v0.12.0 to be merged.

File: docs/contributing/events/kprobes/file_modification.md
Comment (rafaeldtinoco) [inline docs/contributing/events/kprobes/file_modification.md:27]: nit: ` t ` spell.

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [inline pkg/ebpf/c/tracee.bpf.c:4016]: Are you aware that file_modified, or, even, file_update_time, aren't functions called by all existing filesystems, right? Is this by design?

I mean, you're clearly missing file modifications coming from ext2, jfs, vfat, tmpfs, etc...

Comment (rafaeldtinoco) [reply pkg/ebpf/c/tracee.bpf.c:4016]: If that is acceptable (for not at least) then maybe you should document this in the markdown for the event (like you did talking about the LRU map and possibility of sending the event twice).

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [inline pkg/ebpf/c/tracee.bpf.c:4011]: I can only see file_modified being called by:

- btrfs
- cifs
- ext4
- f2fs
- fuse
- xfs

And can see file_update_time being called by:

- 9p
- afs
- btrfs
- ceph
- ext2 (e.g. only when DAX is enabled)
- ext4
- f2fs
- fuse
- gfs2
- kernfs
- ntfs
- ntfs3
- orangefs
- ubifs
- xfs
- zonefs
- mm (for write faults of mmap files)

Before merging, are you aware of that? I believe some filesystems just mark the inode as dirty and rely in the super block operations struct to commit (instead of relying in the inode operations struct like the more modern ones).

File: tests/e2e-instrumentation-test.sh
Comment (rafaeldtinoco) [inline tests/e2e-instrumentation-test.sh:1]: Eventually we will have to normalize this file against all e2e tests being added here (or unify these tests with the work being done by @geyslan in the `test-integration` logic).

**Review by rafaeldtinoco — APPROVED** [review]
LGTM but Id like a clarification regarding of supported filesystems in the event MD, for example, or a discussion about possible other probes so this could be generic for all filesystems (if there is a way).

**Review by rafaeldtinoco — CHANGES_REQUESTED** [review]
@roikol it needs rebasing, and I asked Yaniv to take another look. Maybe you can check with him about the probe points and supported filesystems as well.

**Comment by roikol** [comment]
closing now. will try different approach

**Pull request closed without merging**
